e2e-tests/IIoTPlatform-E2E-Tests/Twin/TwinTestContext.cs

```diff
@@ -0,0 +1,101 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace IIoTPlatform_E2E_Tests.Twin {
+    using IIoTPlatform_E2E_Tests.TestExtensions;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using Xunit;
+
+    public class TwinTestContext : IIoTPlatformTestContext {
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TwinTestContext"/> class.
+        /// Used for preparation executed once before any tests of the collection are started.
+        /// </summary>
+        public TwinTestContext() : base() {
+            PrepareTestEnvironment();
```

Comment:
I would suggest to move call to `PrepareTestEnvironment();` to the constructor of test suite, where `TwinTestContext` will be injected. As it is at the moment, the constructor of `TwinTestContext` does too much work which will also take a considerable amount of time to finish.

Reply:
Having it here ensures that this is executed only once before the test collection is started. This refactoring was needed because there will be more classes included in this test collection, so it will decrease overall test execution time (by doing only one preparation and cleanup).

Reply:
Wouldn't it also be executed once if you call it from the constructor of `TwinBrowseTestTheory` ?

Reply:
One other aspect of the current setup that bothers me is that during debugging session you would have to wait for all this code in constructor of `TwinTestContext` to execute before the execution reaches constructor of `TwinBrowseTestTheory`. That can result is a delay of several minutes before code execution reaches your test suite, which can be very surprising for somebody that is not familiar with this code.

Reply:
Moving the setup to the TwinBrowseTestTheory class will increase the execution time because we will have at least two more classes for the twin tests (TwinCallTestTheory, TwinReadNodeTestTheory) so in my opinion this is better. Also having one setup for the whole collection is a common practice in xunit and shouldn't be a big surprise for a developer.

Reply:
The constructor of the test class is not executed once before it's test methods, but before each test method call. In that case you would have the ~1 min preparation before each test. You can solve this with a class fixture, or a context of a testcollection. You won't be able to follow these in the test explorer during debugging, which is a bit annoying, but overall test execution time should decrease dramatically.

Reply:
OK, got it. I kind of forgot about the suite being instantiated once per each test case. And I didn't know that the same fixture object will be shared among the test suites as well, good to know. Thanks for clarification.